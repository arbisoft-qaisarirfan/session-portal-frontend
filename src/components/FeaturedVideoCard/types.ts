export type FeaturedVideoCardProps = {
  isVisible?: boolean;
  className?: string;
  event_time: string;
  thumbnail?: string;
  title: string;
  description: string;
  video_duration: number;
  workstream_id: string;
  width?: string;
};
