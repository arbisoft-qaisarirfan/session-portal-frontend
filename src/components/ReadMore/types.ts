export interface ReadMoreProps {
  amountOfWords?: number;
  className?: string;
  showLessText: string;
  showMoreText: string;
  text: string;
}
