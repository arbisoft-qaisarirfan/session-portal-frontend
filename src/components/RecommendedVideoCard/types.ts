export interface RecommendedVideoCardProps {
  className?: string;
  date: string;
  duration: string;
  imgUrl: string;
  onClick?: VoidFunction;
  organizer?: string;
  title: string;
  width?: string;
}
