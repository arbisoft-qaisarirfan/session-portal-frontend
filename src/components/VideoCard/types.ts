export type VideoCardProps = {
  className?: string;
  data: {
    event_time: string;
    organizer: string;
    thumbnail: string;
    title: string;
    video_duration: string;
  };
  onClick?: VoidFunction;
  width?: string;
};
