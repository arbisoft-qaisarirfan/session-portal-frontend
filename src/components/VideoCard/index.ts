import VideoCard from "./videoCard";

export * from "./types";

export default VideoCard;
