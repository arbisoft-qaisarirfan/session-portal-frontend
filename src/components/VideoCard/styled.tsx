import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import { cardContentClasses } from "@mui/material/CardContent";
import { skeletonClasses } from "@mui/material/Skeleton";
import { styled, css, alpha } from "@mui/material/styles";
import { typographyClasses } from "@mui/material/Typography";

import { shouldForwardProp } from "@/utils/styleUtils";

export const VideoCardContainer = styled(Card, {
  name: "VideoCardContainer",
  shouldForwardProp,
})<{ $width: string }>(({ theme, $width }) => {
  return css`
    background: transparent;
    border-radius: unset;
    cursor: pointer;
    display: grid;
    overflow: unset;
    width: ${$width};

    .${cardContentClasses.root} {
      align-items: flex-start;
      display: flex;
      flex-direction: column;
      gap: ${theme.spacing(10 / 8)};
      padding: 0;
      width: 100%;

      :last-child {
        padding-bottom: 0px;
      }

      img {
        border-radius: ${theme.shape.borderRadius + 8}px;
        height: auto;
        max-height: 192px;
        min-height: 192px;
        object-fit: cover;
        width: 100%;
      }

      .${skeletonClasses.root} {
        border-radius: ${theme.shape.borderRadius + 8}px;
        position: absolute;
        top: 0;
        z-index: -1;
        width: 100%;
        height: 100%;
      }

      .video-detail {
        display: flex;
        flex-direction: column;
        gap: 7px;
        width: 100%;

        .${typographyClasses.h1}, .${typographyClasses.h5}, .${typographyClasses.h3} {
          -webkit-box-orient: vertical;
          -webkit-line-clamp: 2;
          color: ${theme.palette.colors.white};
          display: -webkit-box;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .date-time,
        .organizer-name {
          color: ${theme.palette.colors.gray};
          font-size: 14px;
          font-weight: 500;
        }

        .video-description {
          -webkit-box-orient: vertical;
          -webkit-line-clamp: 4;
          color: ${theme.palette.colors.white};
          display: -webkit-box;
          font-size: 18px;
          font-weight: 500;
          overflow: hidden;

          ${theme.breakpoints.down("md")} {
            -webkit-line-clamp: 2;
          }
        }
      }
    }

    &.normal-card {
      .${cardContentClasses.root} {
        .video-detail {
          .${typographyClasses.h3} {
            font-size: 16px;
            font-weight: 600;
          }
        }
      }
    }

    &.related-card {
      display: flex;
      margin-bottom: 20px;
      width: 100%;

      .${cardContentClasses.root} {
        flex-direction: row;
        gap: 10px;

        .image-wrapper {
          border-radius: ${theme.shape.borderRadius + 2}px;
          overflow: hidden;
          width: 113px;

          .video-player {
            border-radius: ${theme.shape.borderRadius + 2}px;
          }

          img {
            border-radius: ${theme.shape.borderRadius + 2}px;
            height: unset;
            max-height: unset;
            min-height: unset;
          }

          .${skeletonClasses.root} {
            border-radius: ${theme.shape.borderRadius + 2}px;
          }
        }

        .video-detail {
          gap: 4px;
          width: calc(100% - 113px);

          .${typographyClasses.h5} {
            font-size: 12px;
            font-weight: 500;
          }

          .organizer-name,
          .date-time {
            color: ${theme.palette.colors.gray};
            font-size: 12px;
          }
        }
      }
    }

    &.featured-card {
      background-color: ${alpha(theme.palette.common.white, 0.06)};
      border-radius: 12px;
      display: flex;
      padding: 30px;

      .${cardContentClasses.root} {
        flex-direction: row;
        gap: ${theme.spacing(2.5)};

        ${theme.breakpoints.down("sm")} {
          flex-direction: column;
        }

        .image-wrapper {
          border-radius: ${theme.shape.borderRadius + 8}px;
          height: auto;
          object-fit: cover;
          width: 400px;

          ${theme.breakpoints.down("md")} {
            width: 300px;
          }

          ${theme.breakpoints.down("sm")} {
            width: 100%;
          }

          img {
            border-radius: ${theme.shape.borderRadius + 8}px;
            height: unset;
            max-height: unset;
            min-height: unset;
          }

          .${skeletonClasses.root} {
            border-radius: ${theme.shape.borderRadius + 8}px;
          }
        }

        .video-detail {
          width: calc(100% - 430px);
          gap: 10px;

          ${theme.breakpoints.down("md")} {
            width: calc(100% - 320px);
          }

          ${theme.breakpoints.down("sm")} {
            width: 100%;
          }

          .${typographyClasses.h1} {
            font-weight: 500;
          }

          .organizer-name,
          .date-time {
            font-size: 18px;
          }
        }
      }
    }

    &.search-card {
      display: flex;

      .${cardContentClasses.root} {
        flex-direction: row;
        gap: ${theme.spacing(2.5)};

        ${theme.breakpoints.down("sm")} {
          flex-direction: column;
        }

        .image-wrapper {
          border-radius: ${theme.shape.borderRadius + 8}px;
          width: 400px;

          ${theme.breakpoints.down("md")} {
            width: 300px;
          }

          ${theme.breakpoints.down("sm")} {
            width: 100%;
          }

          img {
            border-radius: ${theme.shape.borderRadius + 8}px;
            height: unset;
            max-height: unset;
            min-height: unset;
          }

          .${skeletonClasses.root} {
            border-radius: ${theme.shape.borderRadius + 8}px;
          }
        }

        .video-detail {
          width: calc(100% - 430px);
          gap: 10px;

          ${theme.breakpoints.down("md")} {
            width: calc(100% - 320px);
          }

          ${theme.breakpoints.down("sm")} {
            width: 100%;
          }

          .${typographyClasses.h1} {
            font-weight: 500;
          }

          .organizer-name,
          .date-time {
            font-size: 18px;
          }
        }
      }
    }

    // for storybook
    &.custom-video-card {
      .${cardContentClasses.root} {
        .video-detail {
          * {
            color: ${theme.palette.secondary.main};
          }
        }
      }
    }
  `;
});

export const ImageWrapper = styled(Box, {
  name: "ImageWrapper",
})(({ theme }) => {
  return css`
    display: inline-flex;
    position: relative;
    width: 100%;

    .video-player {
      border-radius: ${theme.shape.borderRadius + 8}px;
      height: 100%;
      left: 0;
      object-fit: cover;
      position: absolute;
      top: 0;
      width: 100%;
      z-index: 2;
    }

    .video-duration {
      background-color: ${alpha(theme.palette.common.black, 0.7)};
      border-radius: 2px;
      bottom: 10px;
      color: ${theme.palette.colors.white};
      font-size: 12px;
      padding: 2px 4px;
      position: absolute;
      right: 10px;
      z-index: 1;
    }
  `;
});
