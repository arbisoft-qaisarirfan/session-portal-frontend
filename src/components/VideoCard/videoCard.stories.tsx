import type { Meta, StoryObj } from "@storybook/react";

import { VideoCardProps } from "./types";
import VideoCard from "./videoCard";

const meta: Meta<VideoCardProps> = {
  title: "Components/RecommendedVideoCard",
  component: VideoCard,
  tags: ["autodocs"],
  argTypes: {
    title: {
      control: "text",
      description: "The title of the video",
    },
    workstream_id: {
      control: "text",
      description: "The organizer/creator of the video",
    },
    event_time: {
      control: "text",
      description: "The upload date of the video",
    },
    thumbnail: {
      control: "text",
      description: "The URL of the video thumbnail image",
    },
    tags: {
      control: "object",
      description: "The tags available in the video",
    },
    width: {
      control: "text",
      description: "The width of the card (e.g., '100%', '315px')",
    },
  },
};

export default meta;

type Story = StoryObj<VideoCardProps>;

export const Default: Story = {
  args: {
    title: "Sample Video Title",
    workstream_id: "Arbisoft",
    event_time: "2023-10-01",
    thumbnail: "/assets/images/temp-youtube-logo.webp",
    tags: ["Workshop", "Ollama", "AI"],
  },
};

export const LongTitle: Story = {
  args: {
    ...Default.args,
    title: "This is a very long video title that should be truncated",
    width: "315px",
  },
};
