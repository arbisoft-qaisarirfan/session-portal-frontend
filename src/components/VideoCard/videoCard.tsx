"use client";

import React, { FC } from "react";

import Box from "@mui/material/Box";
import CardContent from "@mui/material/CardContent";
import Skeleton from "@mui/material/Skeleton";
import Typography from "@mui/material/Typography";
import Image from "next/image";

import useNavigation from "@/hooks/useNavigation";
import { convertSecondsToFormattedTime, formatDateTime } from "@/utils/utils";

import { ImageWrapper, VideoCardContainer } from "./styled";
import { VideoCardProps } from "./types";

const VideoCard: FC<VideoCardProps> = ({
  id,
  className,
  event_time,
  thumbnail,
  workstream_id,
  title,
  video_duration,
  width = "315px",
}) => {
  const { navigateTo } = useNavigation();
  const inAppThumbnail = `${process.env.NEXT_PUBLIC_BASE_URL}/${thumbnail}`;

  return (
    <VideoCardContainer
      $width={width}
      className={className}
      data-testid="video-card"
      onClick={() => navigateTo("videoDetail", { id })}
    >
      <CardContent>
        <ImageWrapper>
          <Skeleton width="100%" height={192} variant="rounded" animation="wave" />
          <Image
            alt={title}
            height={192}
            width={315}
            src={thumbnail ? inAppThumbnail : "/assets/images/temp-youtube-logo.webp"}
          />
          <Typography className="video-duration" component="div">
            {convertSecondsToFormattedTime(video_duration)}
          </Typography>
        </ImageWrapper>
        <Box className="video-detail">
          <Typography variant="h3" component="div" title={title}>
            {title}
          </Typography>
          <Typography variant="bodyMedium" className="organizer-name" data-testid="video-card-organizer">
            {workstream_id}
          </Typography>
          <Typography variant="bodyMedium" className="date-time" data-testid="video-card-date-time">
            {formatDateTime(event_time)}
          </Typography>
        </Box>
      </CardContent>
    </VideoCardContainer>
  );
};

export default VideoCard;
