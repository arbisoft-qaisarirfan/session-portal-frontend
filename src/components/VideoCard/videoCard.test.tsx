import { customRender as render, screen } from "@/jest/utils/testUtils";

import { VideoCardProps } from "./types";
import VideoCard from "./videoCard";

const mockProps: VideoCardProps = {
  id: 1,
  className: "custom-class",
  event_time: "2024-10-22T12:00:00Z",
  event_type: "SESSION",
  description: "Sample Video Description",
  thumbnail: "assets/images/temp-youtube-logo.webp",
  title: "Sample Video Title",
  publisher: { id: 1, first_name: "John", last_name: "Doe" },
  tags: ["Workshop", "Ollama", "AI"],
  is_featured: false,
  status: "PUBLISHED",
  workstream_id: "Ikram Ali",
  width: "300px",
};

describe("VideoCard", () => {
  test("should renders the component with provided props", () => {
    render(<VideoCard {...mockProps} />);

    expect(screen.getByText(mockProps.title)).toBeInTheDocument();
    expect(screen.getByTestId("video-card-date-time")).toBeInTheDocument();
  });

  test("should displays default image when imgUrl is not provided", () => {
    render(<VideoCard {...mockProps} />);

    const imgUrl = screen.getByRole("img", { name: mockProps.title }).getAttribute("src") ?? "";

    expect(decodeURIComponent(imgUrl)).toContain("/assets/images/temp-youtube-logo.webp");
  });

  test("should renders the organizer name", () => {
    render(<VideoCard {...mockProps} />);
    expect(screen.getByText(mockProps.title)).toBeInTheDocument();
    expect(screen.getByTestId("video-card-organizer")).toBeInTheDocument();
  });

  test("should matches snapshot", () => {
    const { asFragment } = render(<VideoCard {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
