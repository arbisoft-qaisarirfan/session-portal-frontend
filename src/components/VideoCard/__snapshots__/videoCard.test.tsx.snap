// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root custom-class css-1izo6yh-MuiPaper-root-MuiCard-root-VideoCardContainer-root"
    data-testid="video-card"
    style="--Paper-shadow: var(--mui-shadows-1); --Paper-overlay: var(--mui-overlays-1);"
  >
    <div
      class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-jmpm0q-MuiPaper-root-MuiCard-root-ImageWrapper-root"
        style="--Paper-shadow: var(--mui-shadows-1); --Paper-overlay: var(--mui-overlays-1);"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-rounded MuiSkeleton-wave css-116i1gn-MuiSkeleton-root"
          style="width: 100%; height: 192px;"
        />
        <img
          alt="Sample Video Title"
          data-nimg="1"
          decoding="async"
          height="192"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Flocalhost%3A1234%2Fassets%2Fimages%2Ftemp-youtube-logo.webp&w=640&q=75"
          srcset="/_next/image?url=http%3A%2F%2Flocalhost%3A1234%2Fassets%2Fimages%2Ftemp-youtube-logo.webp&w=384&q=75 1x, /_next/image?url=http%3A%2F%2Flocalhost%3A1234%2Fassets%2Fimages%2Ftemp-youtube-logo.webp&w=640&q=75 2x"
          style="color: transparent;"
          width="315"
        />
        <div
          class="MuiTypography-root MuiTypography-bodyMedium video-duration css-149i8z5-MuiTypography-root"
        >
          30:00
        </div>
      </div>
      <div
        class="video-detail MuiBox-root css-0"
      >
        <div
          class="MuiTypography-root MuiTypography-h3 css-1fho9p8-MuiTypography-root"
          title="Sample Video Title"
        >
          Sample Video Title
        </div>
        <p
          class="MuiTypography-root MuiTypography-bodyMedium organizer-name css-149i8z5-MuiTypography-root"
          data-testid="video-card-organizer"
        >
          Ikram Ali
        </p>
        <p
          class="MuiTypography-root MuiTypography-bodyMedium date-time css-149i8z5-MuiTypography-root"
          data-testid="video-card-date-time"
        >
          Oct 22, 2024
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
